crates/turborepo-auth/src/auth/login.rs

```diff
@@ -0,0 +1,334 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+pub use error::Error;
+use reqwest::Url;
+use tokio::sync::OnceCell;
+use tracing::warn;
+use turborepo_api_client::Client;
+use turborepo_ui::{start_spinner, BOLD, UI};
+
+use crate::{error, server::LoginServer, ui};
+
+const DEFAULT_HOST_NAME: &str = "127.0.0.1";
+const DEFAULT_PORT: u16 = 9789;
+
+/// Login writes a token to disk at token_path. If a token is already present,
+/// we do not overwrite it and instead log that we found an existing token.
+pub async fn login<F>(
+    api_client: &impl Client,
+    ui: &UI,
+    existing_token: Option<&str>,
+    mut set_token: F,
+    login_url_configuration: &str,
+    login_server: &impl LoginServer,
+) -> Result<()>
+where
+    F: FnMut(&str) -> Result<()>,
+{
+    // Check if token exists first.
+    if let Some(token) = existing_token {
+        if let Ok(response) = api_client.get_user(&token).await {
+            println!("{}", ui.apply(BOLD.apply_to("Existing token found!")));
+            ui::print_cli_authorized(&response.user.email, ui);
+            return Ok(());
+        }
+    }
+
+    let redirect_url = format!("http://{DEFAULT_HOST_NAME}:{DEFAULT_PORT}");
+    let mut login_url = Url::parse(login_url_configuration)?;
+
+    login_url
+        .path_segments_mut()
+        .map_err(|_: ()| Error::LoginUrlCannotBeABase {
+            value: login_url_configuration.to_string(),
+        })?
+        .extend(["turborepo", "token"]);
+
+    login_url
+        .query_pairs_mut()
+        .append_pair("redirect_uri", &redirect_url);
+
+    println!(">>> Opening browser to {login_url}");
+    let spinner = start_spinner("Waiting for your authorization...");
+
+    // Try to open browser for auth confirmation.
+    {
```

Comment:
Why's this in a block?

Reply:
remnant from when I was trying to understand the borrow checker and/or get around it, should be gone now